### ISSUE TITLE
PDK compiler could be packaged into SDK?

### ISSUE THREAD

**Issue by benwiley4000** [extism/extism#516]
Hi, I'm new to this project but it sounds really promising. I'm not aware of other projects that seem so well suited to supporting arbitrary user code in, for my case, a web app. I investigated iframes first but there didn't seem to be a great way to send and receive messages with an arbitrarily defined iframes, while restricting access to local storage, indexeddb, etc.

My main goal with the plugin system I'd like to add to my app, is to allow users to quickly write a simple JavaScript (or other) function and upload it. I would love to avoid making users download the PDK to test their stuff. I wondered if it could be possible to have (as part of the SDK or as separate packages) the plugin compilers, so a user could upload some JavaScript in the browser, and it instantly gets compiled to wasm and then installed as a plugin.

**Comment by benwiley4000** [comment]
Ok I'll check that out thank you.

I have a question, I know some of the API is considered unstable and will
change. Should I expect that plugins that were previously compiled to wasm
with the PDK might stop working in a newer version of Extism ?

**Comment by nilslice** [comment]
@benwiley4000 - 

This is actually a great time to start an Extism project, as we're moving toward v1.0 by end of the year. The PDK versioning should retain compatibility for quite a long time, once this PR is complete: https://github.com/extism/extism/issues/504

We also have a longer-term compatibility + versioning scheme, which has not yet been implemented. However, if you'd like to take a look and consider providing feedback, we'd appreciate it!https://github.com/extism/proposals/blob/main/EIP-005-extism-versioning-system.md

In any case, I would keep a close eye on https://github.com/extism/extism/issues/504. The crux of the current state is that:
- plugins built before that PR will need to be re-compiled
- plugins built after that PR will be compatible for a long time

Also, follow the #development channel in our Discord too, where we'll be active about this topic. https://extism.org/discord

**Comment by benwiley4000** [comment]
Alright thanks so much for that info!

Ben

**Comment by benwiley4000** [comment]
FYI for now I am using a system based in iframes. I was able to get it to work with an iframe serving a separate html file on my own domain, with sandbox="allow-scripts" and a restrictive CSP meta tag in the html file, which accepts only script evals. I like this solution because it gives users access to the Web Audio API - my application is for audio.

However I am going to keep an eye on Extism, I think it has a lot of potential especially because it supports so many languages.

**Comment by nilslice** [comment]
Going to close this for now, but please feel free to re-open if you visit this again

**Comment by morgante** [comment]
FYI we also would like to have a compiler packaged into the SDK.

**Comment by nilslice** [comment]
@morgante - hey! interesting -- would you elaborate about why and which SDK?

**Comment by morgante** [comment]
We ship a [CLI](https://docs.grit.io/getting-started/cli-quickstart) to our users and have historically had problems with any dependencies we introduce—the CLI itself is a static binary.

We'd like to embed arbitrary JavaScript or AssemblyScript code into the GritQL language, but without requiring additional dependencies.

Ideally it would be JavaScript PDK + Rust SDK.

**Comment by bhelx** [comment]
@morgante Our javascript PDK is a rust project. There is [a CLI crate](https://github.com/extism/js-pdk/tree/main/crates/cli) but I assume you'd be able to use it as a library. That's not something we explicitly support but I'd be happy to help you figure it out. All the pieces are in rust (with the exception of the quickjs code which is a c project wrapped in rust).

**Comment by bhelx** [comment]
Note you may need to kind of copy some of the stuff the CLI crate in that project is doing. But it's not much code. We could work on exposing it as a library and supporting it if it works for you.

**Comment by nilslice** [comment]
oh, @morgante, sorry if you were also the person I was talking to on Discord! didn't connect the `grit` dots :) 

let us know if the JS PDK works for you as a library.. if you end up needing something specific, please open issues on that repo!

**Comment by morgante** [comment]
Thanks! It doesn't look like you have it published though? I think we need a `lib.rs` to be able to import it. Obviously I could copy and paste from the CLI code, but I'd rather avoid forking.

**Comment by bhelx** [comment]
@morgante no, i don't think we've published it as a crate. I don't think it's going to just work as is. If you could just pull it from github and do some experimentation to confirm it's possible, I'd be happy to take what you learned and change the codebase to support it. I just want to make sure it works for you first. Of course let me know if you get stuck or need help. Discord is the best place to reach out and i'm available anytime.
